### PR TITLE
feat(dashboards): Add categorical bar chart widget type

### PR DESCRIPTION
**Pull request by gggritso**
Adds a new **Categorical Bar Chart** widget type to Dashboards that allows users to visualize data with categorical X-axes (e.g., browser, country, transaction) instead of time-series data.

**e.g.,**
Screenshot 2026-02-05 at 12 33 30 PM

The gist of the PR is these changes:

1. Declaring `CATEGORICAL_BAR` as a new kind of widget. It's enabled for most datasets, and there's a new per-dataset field for the default X-axis selection.
2. Adding very many conditionals to make sure that categorical bars load data from the right place, show the right UI (yes to X Axis selection, no to Group By)
3. Shenanigans with state. A categorical bar chart is basically a table with two columns: one column, and one aggregate (e.g., "transaction" column and "p95(span.duration)" aggregate. To keep things simple, I opted to keep this in the Widget Builder's "fields" state. For categorical bars, that state will hold always _exactly two_ fields, one column and one aggregate. This allows clean re-use of data loading that tables use. To make managing this easier, I've added two new dispatch actions to specifically set the aggregate and the column.

There's also a conversion function to transform incoming table data from `/events/` into categorical series that the bar chart likes, and some small bug fixes. Also, this type used to be called "categorical series" so you'll see me updating some of that code.

Closes DAIN-1182, contributes to DAIN-722.

### CODE REVIEW

File: static/app/views/dashboards/datasetConfig/traceMetrics.tsx [hunk 1]
```diff
@@ -188,6 +188,7 @@ export const TraceMetricsConfig: DatasetConfig<
   EventsTimeSeriesResponse,
   EventsTableData
 > = {
+  defaultCategoryField: '',
```
[the remark this answers is not in the document]
Comment (narsaynorath) [reply static/app/views/dashboards/datasetConfig/traceMetrics.tsx]: Yeah metrics are a little weird for this one since these fields load in adhoc and are scoped to a metric. I'd have to check what values are automatically specced to be included, but we can keep this for now since it's feature flagged

File: static/app/views/dashboards/widgetCard/chart.tsx [hunk 2]
```diff
@@ -710,19 +713,58 @@ function BigNumberComponent({
...
-  if (hasCategoricalBarCharts) {
+  if (!hasCategoricalBarCharts) {
+    return null;
+  }
+
+  if (loading || !tableResults?.[0]) {
+    return <LoadingPlaceholder />;
+  }
+
+  // Transform table data to categorical series format
```
Comment (narsaynorath) [inline static/app/views/dashboards/widgetCard/chart.tsx]: Small nit, but I suppose this comment should move down to the `transformTableToCategoricalSeries` call. In this case we could also get rid of it, I don't think it'll add much to understanding the data flow

File: static/app/views/dashboards/datasetConfig/errors.tsx [hunk 3]
```diff
@@ -96,6 +96,7 @@ export const ErrorsConfig: DatasetConfig<
   EventsStats | MultiSeriesEventsStats | GroupedMultiSeriesEventsStats,
   TableData | EventsTableData
 > = {
+  defaultCategoryField: 'transaction',
```
Comment (nikkikapadia) [inline static/app/views/dashboards/datasetConfig/errors.tsx]: nit: wondering if it's worth having this as `title` instead since that's more commonly used in the errors dataset

File: static/app/views/dashboards/widgetBuilder/components/visualize/index.tsx [hunk 4]
```diff
@@ -357,13 +359,58 @@ function Visualize({error, setError}: VisualizeProps) {
...
+    : null;
+
+  const fields = isCategoricalBarWidget
+    ? aggregateFields
+    : usesYAxisState
+      ? state.yAxis
+      : state.fields;
+
+  const canHaveAlias =
+    !isTimeSeriesWidget && !isBigNumberWidget && !isCategoricalBarWidget;
```
Comment (nikkikapadia) [inline static/app/views/dashboards/widgetBuilder/components/visualize/index.tsx]: can this just be 
```suggestion
  const canHaveAlias = isTableWidget;
```
?

File: static/app/views/dashboards/widgetBuilder/hooks/useWidgetBuilderState.tsx [hunk 5]
```diff
@@ -109,6 +112,14 @@ type WidgetAction =
   | {
       payload: TraceMetric | undefined;
       type: typeof BuilderStateAction.SET_TRACE_METRIC;
+    }
+  | {
+      payload: string;
+      type: typeof BuilderStateAction.SET_CATEGORICAL_X_AXIS;
+    }
+  | {
```
Comment (nikkikapadia) [inline static/app/views/dashboards/widgetBuilder/hooks/useWidgetBuilderState.tsx:120]: why do we have the `X_AXIS` as a string but the `AGGREGATE`  as a Column?

File: static/app/views/dashboards/widgetBuilder/components/widgetPreview.tsx [hunk 6]
```diff
@@ -131,7 +131,7 @@ function WidgetPreview({
       // dashboard state to be added
       onWidgetSplitDecision={() => {}}
       // onWidgetSplitDecision={onWidgetSplitDecision}
-
+      tableItemLimit={widget.limit}
```
Comment (nikkikapadia) [inline static/app/views/dashboards/widgetBuilder/components/widgetPreview.tsx:134]: are we going to add a limit for how many x-axis items we show in the widget builder that users can set? 👀  idk how much use it would be but i'd imagine seeing a BUNCH of bars could become a lot.

**Review by nikkikapadia — COMMENTED** [review]
just some nits, comments and questions before I approve. Functionality wise it seems to be looking good! I haven't tested edge cases but I'll test those before i approve.

File: static/app/views/dashboards/widgetBuilder/hooks/useWidgetBuilderState.tsx (hunk 5 above)
Comment (narsaynorath) [reply static/app/views/dashboards/widgetBuilder/hooks/useWidgetBuilderState.tsx:120]: I think with group bys, you _know_ the important bit is just the field name whereas aggregates have to use that complicated format to do all its manipulation of the arguments and stuff so it's easier to deal with aggregates in that way

I think it's okay if we just do a string since down below it forces it into the column format

File: static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+  const {state, dispatch} = useWidgetBuilderContext();
+  const source = useDashboardWidgetSource();
+  const isEditing = useIsEditingWidget();
+
+  const tags: TagCollection = useTags();
+
+  // Only use string tags for categorical X-axis (numeric values don't make good
+  // categories). This has a major caveat that _some_ numerical tags like HTTP
+  // response status _are_ good for grouping, so we may want to allow that.
+  const {tags: stringSpanTags, isLoading: isLoadingSpanTags} = useTraceItemTags('string');
```
Comment (narsaynorath) [inline static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx]: This might be annoying and should possibly be punted to something in the dataset config, but there's some logic to sprinkle in hidden fields and hide some values for the group by, which this is very similar to

https://github.com/getsentry/sentry/blob/bcefb78594a54bbb213529374995e007bee2de22/static/app/views/dashboards/widgetBuilder/components/groupBySelector.tsx#L35

Let me know if you want to do this in your PR here or follow it up since this is all behind a feature flag, this isn't a blocker

File: static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  const isLoading = isEAPDataset && isLoadingSpanTags;
+
+  const fieldOptions = useMemo(() => {
+    // For EAP, use the EAP-style tags and format them directly, we've already
+    // narrowed down to just string tags
+    if (isEAPDataset) {
+      return Object.values(stringSpanTags).map(tag => ({
+        label: prettifyTagKey(tag.name),
+        value: tag.key,
+        trailingItems: () => <TypeBadge kind={FieldKind.FIELD} />,
```
Comment (narsaynorath) [inline static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx]: I don't see the trailing items showing up for me 🤔 

Image

File: static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+      type: BuilderStateAction.SET_CATEGORICAL_X_AXIS,
+      payload: String(option.value),
+    });
+
+    trackAnalytics('dashboards_views.widget_builder.change', {
+      builder_version: WidgetBuilderVersion.SLIDEOUT,
+      field: 'xAxis',
+      from: source,
+      new_widget: !isEditing,
+      value: '',
```
Comment (narsaynorath) [inline static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx]: I think you mean to plumb the `value` into here?

**Review by narsaynorath — APPROVED** [review]
Some small comments, but there are some that need to be addressed before GA, mainly the hidden fields thing. That should actually go into the dataset config, not entirely sure why I didn't put it there earlier on, but regardless for this PR or short term, we can just copy the logic over to make sure stuff is hidden right for the EAP tags

File: static/app/views/dashboards/datasetConfig/traceMetrics.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (gggritso) [reply static/app/views/dashboards/datasetConfig/traceMetrics.tsx]: It won't be feature flagged for long though! Is there a default field that makes sense? If not, I might just update the UI to hide the preview for metrics widgets until a valid X axis is chosen 🤔

File: static/app/views/dashboards/datasetConfig/errors.tsx (hunk 3 above)
Comment (gggritso) [reply static/app/views/dashboards/datasetConfig/errors.tsx]: Fixed!

File: static/app/views/dashboards/widgetBuilder/components/visualize/index.tsx (hunk 4 above)
Comment (gggritso) [reply static/app/views/dashboards/widgetBuilder/components/visualize/index.tsx]: Yep, fixed!

File: static/app/views/dashboards/datasetConfig/errors.tsx (hunk 3 above)
Comment (gggritso) [reply static/app/views/dashboards/datasetConfig/errors.tsx]: Fixed!

File: static/app/views/dashboards/widgetBuilder/components/widgetPreview.tsx (hunk 6 above)
Comment (gggritso) [reply static/app/views/dashboards/widgetBuilder/components/widgetPreview.tsx:134]: This is a good question, I'm not sure yet! I'll add that to the list of things to decide before GA 🙏🏻

File: static/app/views/dashboards/widgetCard/chart.tsx (hunk 2 above)
Comment (gggritso) [reply static/app/views/dashboards/widgetCard/chart.tsx]: Done, thanks!

File: static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx (hunk 7 above)
Comment (gggritso) [reply static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx]: I copypasta'd the hidden tag logic for now 👍🏻

File: static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx (hunk 9 above)
Comment (gggritso) [reply static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx]: I swear that the value was mysteriously missing for other drop-down, so I omitted it here, but ... it's a new analytic, so I don't see why it should be missing. Added!

File: static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx (hunk 8 above)
Comment (gggritso) [reply static/app/views/dashboards/widgetBuilder/components/xAxisSelector.tsx]: Fixed, thanks!

**Review by nikkikapadia — APPROVED** [review]
